_posts/2020-12-10-webpack-vs-sprockets.markdown

```diff
@@ -0,0 +1,313 @@
+
+```
+app
+  |_ javascript
+    |_ packs
+      |_ application.js
+    |_ src
+      |_ some_javascript.js
+    |_ stylesheets
+      |_ application.css
+    |_ images
+      |_ logo.jpg
+      |_ test.png
+    |_ fonts
+      |_ myfont.ttf
+```
+
+And your `application.js` would look like this:
+
+```js
+// packs/application.js
+
+import '../stylesheets/application.css'
+require.context('../images', true) 
+require.context('../fonts', true)
+
+// and now your JavaScript ...
+import SomeModule from '../src/some_javascript'
+```
+
+You can rename the `source_path` to `assets` and use `javascript` instead of `src` and it will look pretty similar to the Assets Pipeline!
+
+## Pros And Cons
+
+So, what should you use? Sprockets? Webpack? Both? As always, the answer is... it depends. By default, since Rails 6, it uses both: Webpack for JavaScript, Sprocket for any other asset. According to your needs, you may want to use one or the other or change which assets you handle with each of them. You can even handle assets of the same type using both solutions at the same time!
+
+Let's do a quick comparison:
+
+### Sprockets Pros
+- As Rails developers, we are really used to how the asset pipeline works and many Rails specific guides may reference this instead of Webpacker (mainly older guides before the Rails 6 release)
+- You can use assets from gems AND from node_modules
+
+### Sprockets Cons
+- There are no plans for make it work with modern JavaScript features ([According to DHH](https://discuss.rubyonrails.org/t/webpacker-presents-a-more-difficult-oob-experience-for-js-sprinkles-than-sprockets-did/75345/9))
+- Bloats the JavaScript global scope
+- It's not the default for JavaScript, so new guides will be focused on Webpack
+
+### Webpack Pros
+- Gives you access to all modern JavaScript features
+- Gives you access to a lot of plugins to handle the compilation process
+- Easy setup for JavaScript frameworks (like React, Vue or Stimulus) when creating Rails apps
+- Does not populate the global scope unless we are explicit
+- Includes an auto-reloader during development
+- There are many resources for Webpack in general (not specific to Rails) that also apply
+
+### Webpack Cons
+- You can (and should) import CSS files (and other asset types) inside JavaScript files, which may be confusing at first
+- From my experience, you can't read assets from gems (documentation states that you can use assets from Rails Engines, but I was not able to set that up)
+
+## Conclusion
+
+After this comparison, I think the general approach of using Webpack for JavaScript files and Sprockets for the rest is the way to go for now. It enables JavaScript modern features using Webpack but leaves the other assets to be handled by Sprockets so the learning curve is not as pronounced for developers used to the Assets Pipeline. And at the same time it's also capable of handling all the assets types using Webpack for developers experiences with Webpack that comes to Rails.
```

Comment:
```suggestion
After this comparison, I think the general approach of using Webpack for JavaScript files and Sprockets for the rest is the way to go for now. It enables JavaScript's modern features using Webpack but leaves the other assets to be handled by Sprockets so the learning curve is not as pronounced for developers used to the Asset Pipeline. And at the same time it's also capable of handling all the asset types using Webpack for developers experiences with Webpack that comes to Rails.
```

Maybe: "However, for developers used to using Webpack, it may be an easier alternative to use Webpack for the handling of all assets.